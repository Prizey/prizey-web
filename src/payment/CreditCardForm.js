import React, { useState } from 'react'
import {
  CardNumberElement,
  CardExpiryElement,
  CardCVCElement,
  injectStripe,
} from 'react-stripe-elements'
import { Grid, Button, CircularProgress } from '@material-ui/core'
import 'react-credit-cards/es/styles-compiled.css'

import ErrorComponent from 'design/Error/Error'
import StripeField from 'design/StripeField'

export const handleSubmit = ({
  setError,
  stripe,
  create,
  purchase,
}) => async evt => {
  evt.preventDefault()

  try {
    const result = await stripe.createToken({ type: 'card' })
    if (result.error) {
      setError(result.error.message)
    } else {
      create({
        credit_card_token: result.token.id,
        purchase_option_id: purchase.id,
      })
    }
  } catch (err) {
    setError(err.message)
  }

  return false
}

export const CreditCardComponent = ({ classes, creating, error, ...props }) => {
  const [stripeError, setError] = useState(null)
  return (
    <form onSubmit={handleSubmit({ setError, ...props })}>
      {/* <CardNumberElement /> */}
      <StripeField label="Card Number" component={CardNumberElement} />

      <Grid container spacing={24}>
        <Grid item xs>
          {/* <CardExpiryElement /> */}
          <StripeField label="Expiry (MM / YY)" component={CardExpiryElement} />
        </Grid>
        <Grid item xs>
          {/* <CardCVCElement /> */}
          <StripeField label="CVC" component={CardCVCElement} />
        </Grid>
      </Grid>

      {creating && <CircularProgress color="primary" size={36} />}
      {error && <ErrorComponent>{error}</ErrorComponent>}
      {stripeError && <ErrorComponent>{stripeError}</ErrorComponent>}
      <Button variant="contained" color="primary" type="submit" fullWidth>
        PAY
      </Button>
    </form>
  )
}

export default injectStripe(CreditCardComponent)
