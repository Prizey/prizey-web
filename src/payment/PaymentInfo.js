import React from 'react'
import { New, List } from 'croods'
import { StripeProvider, Elements } from 'react-stripe-elements'
import { Redirect } from '@reach/router'
import { Typography } from '@material-ui/core'
import { withStyles } from '@material-ui/core/styles'
import Layout from 'design/Layout/Layout'

import GoBack from 'design/GoBack/GoBack'
import CreditCardForm from './CreditCardForm'

import getPurchaseDetail from './purchaseDescription'

const styles = theme => ({
  icon: {
    float: 'left',
    height: theme.spacing.md,
  },
  purchase: {
    marginBottom: theme.spacing.md,
  },
})

export const afterPurchase = navigate => () => navigate('/')

export const PaymentInfoComponent = withStyles(styles)(
  ({ navigate, classes, purchase }) => (
    <StripeProvider apiKey={process.env.REACT_APP_STRIPE_TOKEN}>
      <Layout leftIcon={<GoBack to="/buy-diamonds" />}>
        <Typography align="center" variant="h5" className={classes.purchase}>
          <img
            src="/icons/diamond.png"
            alt="diamond"
            className={classes.icon}
          />
          {getPurchaseDetail(purchase.ticketAmount, purchase.price)}
        </Typography>

        <New
          name="payments"
          render={props => (
            <Elements>
              <CreditCardForm purchase={purchase} {...props} />
            </Elements>
          )}
          afterCreate={afterPurchase(navigate)}
        />
      </Layout>
    </StripeProvider>
  ),
)

export const reducePurchaseState = (list, id) =>
  list.find(item => item.id === parseInt(id, 10))

export default ({ navigate, currentUser, purchaseId }) =>
  currentUser ? (
    <List
      name="purchaseOptions"
      path="/purchase_options"
      render={purchases => (
        <PaymentInfoComponent
          navigate={navigate}
          purchase={reducePurchaseState(purchases, purchaseId)}
        />
      )}
    />
  ) : (
    <Redirect to={`/sign-in?next=/payment/${purchaseId}`} noThrow />
  )
