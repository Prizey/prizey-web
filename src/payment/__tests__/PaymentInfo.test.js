import React from 'react'
import renderer from 'react-test-renderer'
import PaymentInfo, { afterPurchase } from '../PaymentInfo'
import { handleSubmit } from '../CreditCardForm'

const mockList = [
  { id: 1, price: '10.0', ticketAmount: 10 },
  { id: 2, price: '25.0', ticketAmount: 25 },
  { id: 3, price: '50.0', ticketAmount: 50 },
]

window.Stripe = () => {}

jest.mock('../../design/Layout/RegisterPageView')

jest.mock('react-stripe-elements', () => ({
  CardCVCElement: props => <input {...props} />,
  CardExpiryElement: props => <input {...props} />,
  CardNumberElement: props => <input {...props} />,
  Elements: props => <div {...props} />,
  StripeProvider: ({ children }) => <div>{children}</div>,
  injectStripe: Component => Component,
}))

jest.mock('croods', () => ({
  List: props => (
    <div {...props}>
      {props.children} - Render: {props.render(mockList)}
    </div>
  ),
  New: props => (
    <div {...props}>
      New - <div>render - {props.render(props)}</div>
    </div>
  ),
}))

it('renders correctly', () => {
  const tree = renderer
    .create(<PaymentInfo currentUser purchaseId={1} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('redirects correctly', () => {
  const tree = renderer.create(<PaymentInfo />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe('test the actions after payment', () => {
  it('redirects after purchase and update the tickets', () => {
    const params = {
      currentUser: { id: 1, tickets: 0 },
      navigate: jest.fn(),
      purchase: { ticketAmount: 10 },
      setCurrentUser: jest.fn(),
    }
    afterPurchase(params)()

    expect(params.setCurrentUser).toHaveBeenCalledWith({
      id: 1,
      tickets: 10,
    })
    expect(params.navigate).toHaveBeenCalledWith('/game')
  })
})

it('handle the submit process on sucess', async () => {
  const params = {
    create: jest.fn(),
    purchase: { id: 1 },
    setError: jest.fn(),
    stripe: { createToken: () => Promise.resolve({ token: { id: '101010' } }) },
  }

  await handleSubmit(params)({ preventDefault: jest.fn() })

  expect(params.create).toHaveBeenCalledWith({
    credit_card_token: '101010',
    purchase_option_id: 1,
  })
})

it('handle the submit process on error', async () => {
  const params = {
    create: jest.fn(),
    purchase: { id: 1 },
    setError: jest.fn(),
    stripe: {
      createToken: () =>
        Promise.resolve({ error: { message: 'invalid number' } }),
    },
  }

  await handleSubmit(params)({ preventDefault: jest.fn() })

  expect(params.setError).toHaveBeenCalledWith('invalid number')
})

it('handle the submit process with a crash', async () => {
  const params = {
    create: jest.fn(),
    purchase: { id: 1 },
    setError: jest.fn(),
    stripe: {
      createToken: () => {
        throw new Error('network error')
      },
    },
  }

  await handleSubmit(params)({ preventDefault: jest.fn() })

  expect(params.setError).toHaveBeenCalledWith('network error')
})
