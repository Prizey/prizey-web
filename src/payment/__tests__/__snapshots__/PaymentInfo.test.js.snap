// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get the purchase details 1`] = `"10 Diamonds = $ 10.00"`;

exports[`redirects correctly 1`] = `null`;

exports[`renders correctly 1`] = `
<div
  name="purchaseOptions"
  path="/purchase_options"
  render={[Function]}
>
   - Render: 
  <div
    apiKey="pk_test_iZ0Zhbe6LCwVBFX1B7b0qhWq00eFCM2s7B"
  >
    <div
      className="Component-root-7"
    >
      <div
        className="Component-header-6"
      >
        <a
          aria-label="Go Back"
          className="MuiButtonBase-root-15 MuiIconButton-root-9 Component-root-8"
          href="/buy-diamonds"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          <span
            className="MuiIconButton-label-14"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-18 MuiSvgIcon-fontSizeLarge-26"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
              />
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-247"
          />
        </a>
      </div>
      <div
        className="Component-container-5"
      >
        <h2
          className="MuiTypography-root-27 MuiTypography-h2-40 MuiTypography-alignCenter-50"
        >
          PRIZEY
        </h2>
        <div
          className="Component-caption-3"
        />
        <div
          className="Component-children-4"
        >
          <h5
            className="MuiTypography-root-27 MuiTypography-h5-43 MuiTypography-alignCenter-50 Component-purchase-2"
          >
            <img
              alt="diamond"
              className="Component-icon-1"
              src="/icons/diamond.png"
            />
            10 Diamonds = $ 10.00
          </h5>
          <div
            afterCreate={[Function]}
            name="payments"
            render={[Function]}
          >
            New - 
            <div>
              render - 
              <div>
                <form
                  onSubmit={[Function]}
                >
                  <fieldset
                    className="MuiFormControl-root-63 MuiFormControl-fullWidth-66"
                  >
                    <label
                      className="MuiFormLabel-root-78 MuiInputLabel-root-67 MuiInputLabel-formControl-72 MuiInputLabel-animated-75"
                      data-shrink={false}
                    >
                      Card Number
                    </label>
                    <div
                      className="MuiInputBase-root-98 MuiInput-root-85 MuiInput-underline-89 MuiInputBase-fullWidth-107 MuiInput-fullWidth-92 MuiInputBase-formControl-99 MuiInput-formControl-86"
                      onClick={[Function]}
                    >
                      <input
                        className="Component-root-115"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        style={
                          Object {
                            "base": Object {
                              "color": "rgba(0, 0, 0, 0.87)",
                              "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                              "fontSize": "14px",
                            },
                            "invalid": Object {
                              "color": "#f44336",
                            },
                          }
                        }
                      />
                    </div>
                    <p
                      className="MuiFormHelperText-root-116"
                    />
                  </fieldset>
                  <div
                    className="MuiGrid-container-124 MuiGrid-spacing-xs-24-148"
                  >
                    <div
                      className="MuiGrid-item-125 MuiGrid-grid-xs-true-152"
                    >
                      <fieldset
                        className="MuiFormControl-root-63 MuiFormControl-fullWidth-66"
                      >
                        <label
                          className="MuiFormLabel-root-78 MuiInputLabel-root-67 MuiInputLabel-formControl-72 MuiInputLabel-animated-75"
                          data-shrink={false}
                        >
                          Expiry (MM / YY)
                        </label>
                        <div
                          className="MuiInputBase-root-98 MuiInput-root-85 MuiInput-underline-89 MuiInputBase-fullWidth-107 MuiInput-fullWidth-92 MuiInputBase-formControl-99 MuiInput-formControl-86"
                          onClick={[Function]}
                        >
                          <input
                            className="Component-root-115"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            style={
                              Object {
                                "base": Object {
                                  "color": "rgba(0, 0, 0, 0.87)",
                                  "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                                  "fontSize": "14px",
                                },
                                "invalid": Object {
                                  "color": "#f44336",
                                },
                              }
                            }
                          />
                        </div>
                        <p
                          className="MuiFormHelperText-root-116"
                        />
                      </fieldset>
                    </div>
                    <div
                      className="MuiGrid-item-125 MuiGrid-grid-xs-true-152"
                    >
                      <fieldset
                        className="MuiFormControl-root-63 MuiFormControl-fullWidth-66"
                      >
                        <label
                          className="MuiFormLabel-root-78 MuiInputLabel-root-67 MuiInputLabel-formControl-72 MuiInputLabel-animated-75"
                          data-shrink={false}
                        >
                          CVC
                        </label>
                        <div
                          className="MuiInputBase-root-98 MuiInput-root-85 MuiInput-underline-89 MuiInputBase-fullWidth-107 MuiInput-fullWidth-92 MuiInputBase-formControl-99 MuiInput-formControl-86"
                          onClick={[Function]}
                        >
                          <input
                            className="Component-root-115"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            style={
                              Object {
                                "base": Object {
                                  "color": "rgba(0, 0, 0, 0.87)",
                                  "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                                  "fontSize": "14px",
                                },
                                "invalid": Object {
                                  "color": "#f44336",
                                },
                              }
                            }
                          />
                        </div>
                        <p
                          className="MuiFormHelperText-root-116"
                        />
                      </fieldset>
                    </div>
                  </div>
                  <button
                    className="MuiButtonBase-root-15 MuiButton-root-221 MuiButton-contained-232 MuiButton-containedPrimary-233 MuiButton-raised-235 MuiButton-raisedPrimary-236 MuiButton-fullWidth-246"
                    disabled={false}
                    onBlur={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="submit"
                  >
                    <span
                      className="MuiButton-label-222"
                    >
                      PAY
                    </span>
                    <span
                      className="MuiTouchRipple-root-247"
                    />
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <p
          className="MuiTypography-root-27 MuiTypography-body1-36 MuiTypography-alignCenter-50"
        />
      </div>
    </div>
  </div>
</div>
`;
